### ISSUE TITLE
Distribute modules into separate libraries that can be imported and registered optionally (separation of concerns).

### ISSUE THREAD

**Issue by EdDuarte** [vokter/vokter-core#4]
This task is a huge undertaking that I intend to get around to very soon. It involves a complete re-do over the original project's scope and a large restructuring of the full codebase.

Below is my current plan, but I might change it as development kicks in. The proposal below always reflects the most recent proposal (this first post is edited to match the current plan), and any further plan changes will be added as comments to this issue.

-----

- **vokter-core** - simplified focus: find changes between two documents (web documents supported by fetching urls) and retrieve diffs back to the developer instantly (**NOT** over time). Uses a single thread (the same thread as the one that calls vokter), so no Listener patterns are required. Includes a plain-text reader (no cleaning or special rules), a document reader for local files, and a stopword file loader (which can be used optionally);

(Every module below is **OPTIONAL** when using vokter-core! vokter-core is the only common library used by all other modules and contains all required interfaces.)

- **vokter-cli** - runs an instance of vokter in a command line process. Contains parameters to register the modules below;

- **vokter-readers** - classes that instruct vokter-core how to interpret and clean the input documents of XML, JSON and HTML formats;

- **vokter-stemmers** - classes that instruct vokter-core how to stem the input documents and keywords. Can be registered.

After that, we move all scheduler logic to vokter-scheduler, which also contains its own set of modules:

- **vokter-scheduler** - integrates Quartz scheduler and runs a memory process for periodic detection and keyword-matching over time, dealing with storage (which, for this scope, only in-memory is provided) of jobs and document snapshots and allowing clients to register a permanent response channel (which, for this scope, corresponds to Listener/Consumer classes or the running console). Can be optionally registered with vokter-cli (which should support scheduler options like job interval), vokter-readers, vokter-stemmers, vokter-scheduler-mongodb and vokter-scheduler-cassandra;

- **vokter-scheduler-mongodb**, **vokter-scheduler-cassandra** and **vokter-scheduler-redis** - integrates MongoDB, Cassandra and Redis respectively as data stores for vokter-scheduler, allowing jobs and document snapshots to persist these between instances.

Finally, we move HTTP server logic to vokter-server, which represents the current repository scope and meta:

- **vokter-server** - deploys vokter-scheduler on a HTTP server and provides a REST API (**WITHOUT** Swagger documentation). Can be optionally registered with vokter-cli, vokter-readers, vokter-stemmers and vokter-scheduler persistence modules;

- **vokter-server-swagger** -  web GUI that can be deployed in parallel with vokter-server while enabling public-access swagger web documentation;

- **vokter-server-dashboard** - a web GUI that can be deployed in parallel with vokter-server that displays all supported operations;

- **vokter-client-java** - deploys a sample HTTP client that writes and reads on a running vokter-server, using its REST API.

Additional Notes:

- I will have to consider which modules will remain in vokter/vokter repository as separate Maven modules (as it currently is) and which ones will be moved into their own repositories. Personally, I would want to move out larger components that heavily affect this repository's insights, creating undesired analytics entropy. One such component is the server web frontends like the swagger docs and the dashboard, which change the repository most common languages from Java to Javascript and CSS. I will update this issue with my final decision on this as we get closer to the final version of these changes;

- the "register" method should have the same paradigm for all its usages, which is adding a new module that actuates simultaneously with previous modules, **NOT** replacing previous modules. With this implicit logic, when registering both vokter-scheduler-mongodb and vokter-scheduler-cassandra, jobs are persisted and synchronized to **BOTH** databases. This allows Vokter to be easily configured to replicate data and optimize queries by maintaining data accessible in both an in-memory database and a disk database. The first registered persistence module becomes the default module used for queries (so users should register an in-memory database first);

- there's two alternative approaches to the vokter-readers module:
  - option 1: vokter-readers could be distributed separately into vokter-reader-xml, vokter-reader-json and vokter-reader-html, or;
  - option 2: vokter-readers could be part of vokter-core, due to their simple nature, and the Reader interface can still be extended by developers to integrate input cleaning to new formats.

- vokter-stopwords, which just contains the same stopword files as the ones distributed by [Snowball](http://snowball.tartarus.org/), should be removed from the project source code, and it's the developer responsibility to download and register the stopword files himself.

- vokter-stemmers correspond to the [Snowball](http://snowball.tartarus.org/) stemmers in Java, so these classes should be imported from Maven Central Repository instead. vokter-stemmers module would still implement classes for each language while implementing the vokter-core Stemmer interface. Perhaps this module should be called "vokter-stemmers-snowball", so as to limit the repository only to Snowball's stemmers.

-----

In summary, a complete restructuring of modules enables Vokter to be used in multiple ways:
- as a library, which can be integrated into any Java project;
- as a CLI app, which allows developers to use vokter and vokter-scheduler functionality without worrying about coding or dependences;
- as a RESTful app, which allows developers to integrate vokter as a service in their infrastructure stack.

**Comment by EdDuarte** [comment]
Original proposal for the new module structure (first post of issue is edited to match the most recent proposal):

- **vokter-core** - simplified focus: find changes between two documents (web documents supported by fetching urls) and retrieve diffs back to the developer instantly (**NOT** over time). Uses a single thread (the same thread as the one that calls vokter), so no Listener patterns are required. Includes a plain-text reader (no cleaning or special rules), a document reader for local files, and a stopword file loader (which can be used optionally);

(Every module below is **OPTIONAL** when using vokter-core! vokter-core is the only common library used by all other modules and contains all required interfaces.)

- **vokter-cli** - runs an instance of vokter in a command line process. Contains parameters to register the modules below;

- **vokter-readers** - classes that instruct vokter-core how to interpret and clean the input documents of XML, JSON and HTML formats;

- **vokter-stemmers** - classes that instruct vokter-core how to stem the input documents and keywords. Can be registered.


After that, we move all scheduler logic to vokter-scheduler, which also contains its own set of module:

- **vokter-scheduler** - integrates Quartz scheduler and runs a memory process for periodic detection and keyword-matching over time, dealing with storage (which, for this scope, only in-memory is provided) of jobs and document snapshots and allowing clients to register a permanent response channel (which, for this scope, corresponds to Listener/Consumer classes or the running console). Can be optionally registered with vokter-cli (which should support scheduler options like job interval), vokter-readers, vokter-stemmers, vokter-scheduler-mongodb and vokter-scheduler-cassandra;

- **vokter-scheduler-mongodb** & **vokter-scheduler-cassandra** - integrates MongoDB and Cassandra respectively as data stores for vokter-scheduler, allowing jobs and document snapshots to persist these between instances.


Finally, we move HTTP server logic to vokter-server, which represents the current repository scope and meta:

- **vokter-server** - deploys vokter-scheduler on a HTTP server and provides a REST API (**WITHOUT** Swagger documentation). Can be optionally registered with vokter-cli, vokter-readers, vokter-stemmers, vokter-scheduler-mongodb and vokter-scheduler-cassandra;

- **vokter-server-swagger** -  web GUI that can be deployed in parallel with vokter-server while enabling public-access swagger web documentation;

- **vokter-server-dashboard** - a web GUI that can be deployed in parallel with vokter-server that displays all supported operations;

- **vokter-client-java** - deploys a sample HTTP client that writes and reads on a running vokter-server, using its REST API.

Additional Notes:

- I have to consider which modules will remain in vokter/vokter repository as separate Maven modules (as it currently is) and which ones will be moved into their own repositories. Personally, I would want to move larger components that heavily affect this repository's insights, creating undesired analytics entropy. One such component is the server web frontends like the swagger docs and the dashboard, which change the repository most common languages from Java to Javascript+CSS. I will update this issue with my final decision on this as we get closer to the final version of these changes;

- when inputting documents to vokter-core, an optional parameter "lang" should exist so that the developer can circumvent the language-detection classifier and force the document to be read in a specific language;

- there's two alternative approaches to the vokter-readers module:
  - option 1: vokter-readers could be distributed separately into vokter-reader-xml, vokter-reader-json and vokter-reader-html, or;
  - option 2: vokter-readers could be part of vokter-core, due to their simple nature, and the Reader interface can still be extended by developers to integrate input cleaning to new formats.

- vokter-stopwords, which just contains the same stopword files as the ones distributed by [Snowball](http://snowball.tartarus.org/), should be removed from the project source code, and it's the developer responsibility to download and register the stopword files himself.

- vokter-stemmers correspond to the [Snowball](http://snowball.tartarus.org/) stemmers in Java, so these classes should be imported from Maven Central Repository instead. vokter-stemmers module would still implement classes for each language while implementing the vokter-core Stemmer interface. Perhaps this module should be called "vokter-stemmers-snowball", so as to limit the repository only to Snowball's stemmers.

**Comment by EdDuarte** [comment]
The additional note related to the introduction of an optional parameter "lang" on subscriptions has been moved to its own issue, #5

**Comment by EdDuarte** [comment]
Added module "vokter-scheduler-redis", which just as the name implies, stores and queries document snapshots and jobs in Redis.

Added the following to "Additional Notes":
- the "register" method should have the same paradigm for all its usages, which is adding a new module that actuates simultaneously with previous modules, **NOT** replacing previous modules. With this implicit logic, when registering both vokter-scheduler-mongodb and vokter-scheduler-cassandra, jobs are persisted and synchronized to BOTH databases. This allows Vokter to be easily configured to replicate data and optimize queries by maintaining data accessible in both an in-memory database and a disk database. The first registered persistence module becomes the default module used for queries (so users should register an in-memory database first).
